src/System/Random/Internal.hs

```diff
@@ -0,0 +1,1010 @@
+
+  -- | Generates a 'Word64' that is uniformly distributed over the entire
+  -- 'Word64' range.
+  --
+  -- The default implementation combines two 'Word32' from 'uniformWord32' into
+  -- one 'Word64'.
+  --
+  -- @since 1.2
+  uniformWord64 :: g -> m Word64
+  uniformWord64 g = do
+    l32 <- uniformWord32 g
+    h32 <- uniformWord32 g
+    pure (unsafeShiftL (fromIntegral h32) 32 .|. fromIntegral l32)
+
+  -- | @uniformShortByteString n g@ generates a 'ShortByteString' of length @n@
+  -- filled with pseudo-random bytes.
+  --
+  -- @since 1.2
+  uniformShortByteString :: Int -> g -> m ShortByteString
+  default uniformShortByteString :: MonadIO m => Int -> g -> m ShortByteString
+  uniformShortByteString n = genShortByteStringIO n . uniformWord64
+  {-# INLINE uniformShortByteString #-}
+
+
+
+-- | This class is designed for stateful RNGs that can be saved as immutable data type.
+--
+-- @since 1.2
+class StatefulGen (MutableGen f m) m => FrozenGen f m where
+  -- | Represents the state of the pseudo-random number generator for use with
+  -- 'thawGen' and 'freezeGen'.
+  --
+  -- @since 1.2
+  type MutableGen f m = (g :: Type) | g -> f
+  -- | Saves the state of the pseudo-random number generator as a frozen 'Seed'.
+  --
+  -- @since 1.2
+  freezeGen :: MutableGen f m -> m f
+  -- | Restores the pseudo-random number generator from its frozen 'Seed'.
+  --
+  -- @since 1.2
+  thawGen :: f -> m (MutableGen f m)
+
+
+data MBA s = MBA (MutableByteArray# s)
+
+
+-- | Efficiently generates a sequence of pseudo-random bytes in a platform
+-- independent manner. The allocated memory is be pinned, so it is safe to use
+-- with FFI calls.
+--
+-- @since 1.2
+genShortByteStringIO :: MonadIO m => Int -> m Word64 -> m ShortByteString
+genShortByteStringIO n0 gen64 = do
+  let !n@(I# n#) = max 0 n0
+      (n64, nrem64) = n `quotRem` 8
+  MBA mba# <-
+    liftIO $
+    IO $ \s# ->
+      case newPinnedByteArray# n# s# of
+        (# s'#, mba# #) -> (# s'#, MBA mba# #)
+  let go i ptr
+        | i < n64 = do
+          w64 <- gen64
+          -- Writing 8 bytes at a time in a Little-endian order gives us
+          -- platform portability
+          liftIO $ runF word64LE w64 ptr
+          go (i + 1) (ptr `plusPtr` 8)
+        | otherwise = return ptr
+  ptr <- go 0 (Ptr (byteArrayContents# (unsafeCoerce# mba#)))
+  when (nrem64 > 0) $ do
+    w64 <- gen64
+    -- In order to not mess up the byte order we write generated Word64 into a
+    -- temporary pointer and then copy only the missing bytes over to the array.
+    -- It is tempting to simply generate as many bytes as we still need using
+    -- smaller generators (eg. uniformWord8), but that would result in
+    -- inconsistent tail when total length is slightly varied.
+    liftIO $ do
+      let goRem64 z i =
+            when (i < nrem64) $ do
+              pokeByteOff ptr i (fromIntegral z :: Word8)
+              goRem64 (z `unsafeShiftR` 8) (i + 1)
```

Comment:
There is usually no reason to use `unsafeShiftR` instead of `shiftR`, when shift is statically known.

Reply:
That's interesting! Could you expand on this a bit more, I'd love to know why is this the case. I am sure the answer is ghc is smart ;) nevertheless I'd like to know how.

Reply:
`shiftR x s` is eventually defined as `if s >= 64 then 0 else unsafeShiftR x s`. So when `s` is statically known and below 64, `shiftR` is replaced by `unsafeShiftR` in Core.

Reply:
That's very nice and good to know about this optimization too. Thanks! I'll update it.